scripts/aws.sh

```diff
@@ -0,0 +1,175 @@
+
+    aws ec2 authorize-security-group-ingress \
+        --group-id $sg_id \
+        --protocol tcp \
+        --port 2222 \
+        --cidr 0.0.0.0/0 \
+        > /dev/null
+
+    aws ec2 authorize-security-group-ingress \
+        --group-id $sg_id \
+        --protocol tcp \
+        --port 80 \
+        --cidr 0.0.0.0/0 \
+        > /dev/null
+
+    aws ec2 authorize-security-group-ingress \
+        --group-id $sg_id \
+        --protocol tcp \
+        --port 443 \
+        --cidr 0.0.0.0/0 \
+        > /dev/null
+
+    aws ec2 authorize-security-group-egress \
+        --group-id $sg_id \
+        --protocol -1 \
+        --port -1 \
+        --cidr 0.0.0.0/0 \
+        > /dev/null
+
+    echo "$sg_id"
+}
+
+create_network_routing() {
+    local vpc_id=$1
+    local subnet_id=$2
+    local owner=$3
+
+    local igw_id=$(aws ec2 create-internet-gateway \
+        --query 'InternetGateway.InternetGatewayId' \
+        --output text)
+
+    aws ec2 create-tags --resources $igw_id --tags "Key=owner,Value=${owner}"
+
+    aws ec2 attach-internet-gateway \
+        --vpc-id $vpc_id \
+        --internet-gateway-id $igw_id \
+        > /dev/null
+
+    local route_table_id=$(aws ec2 create-route-table \
+        --vpc-id $vpc_id \
+        --query 'RouteTable.RouteTableId' \
+        --output text)
+
+    aws ec2 create-tags --resources $route_table_id --tags "Key=owner,Value=${owner}"
+
+    aws ec2 create-route \
+        --route-table-id $route_table_id \
+        --destination-cidr-block 0.0.0.0/0 \
+        --gateway-id $igw_id \
+        > /dev/null
+
+    aws ec2 associate-route-table \
+        --subnet-id $subnet_id \
+        --route-table-id $route_table_id \
+        > /dev/null
+}
+
+create_ec2_instance() {
+    local run_id=$1
+    local subnet_id=$2
+    local sg_id=$3
+    local ssh_public_key=$4
+    local user_data_path=$5
+    local owner=$6
+
+    local ami_id=$(aws ec2 describe-images \
+        --owners $AMI_OWNER \
+        --filters "Name=name,Values=$AMI_NAME" \
+        --query 'sort_by(Images, &CreationDate)[-1].ImageId' \
+        --output text)
+
+    aws ec2 import-key-pair \
+        --key-name "admin-key-pair-${run_id}" \
+        --public-key-material "fileb://${ssh_public_key}"
+
+    local instance_id=$(aws ec2 run-instances \
+        --image-id $ami_id \
+        --instance-type $INSTANCE_TYPE \
+        --subnet-id $subnet_id \
+        --security-group-ids $sg_id \
+        --key-name "admin-key-pair-${run_id}" \
+        --associate-public-ip-address \
+        --user-data "file://${user_data_path}" \
+        --block-device-mappings "[{\"DeviceName\":\"/dev/sda1\",\"Ebs\":{\"VolumeSize\":256}}]" \
+        --query 'Instances[0].InstanceId' \
+        --output text)
+
+    aws ec2 create-tags --resources $instance_id --tags "Key=owner,Value=${owner}"
+
+    aws ec2 wait instance-running --instance-ids $instance_id
+
+    echo "$instance_id"
+}
+
+provision_resources_aws() {
```

Comment:
Can we create resources like vpc and security group, igw, and user key data in parallel by using background task and then do the attachment later with wait command? Otherwise, this will be unfair comparison to Azure because resources are created sequentially while Azure have some resources created in parallel which can make creation faster.